File: XRT/src/Tokenizer.ts
import { SourceLocation } from "./SourceLocation"
import chalk from 'chalk'

enum TokenType {
    KEYWORD,
    IDENTIFIER,
    INVALID_IDENTIFIER,
    LITERAL,
    INVALID_LITERAL,
    OPERATOR,
    PUNCTUATOR,
    COMMENT
}

export interface Token extends SourceLocation {
    type: TokenType;
    value: string;
}

const s_Identifiers: string[] = [
    "namespace",
    "component",
    "ports",
    "registers",
    "implementation",
    "extern",
    "in",
    "out",
    "inout",
    "if",
    "else",
    "for",
    "template",
    "typename",
    "include",
    "using",
].sort()

const regex_Literal = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/
const regex_HexLiteral = /^0x[0-9a-fA-FxX_']+$/
const regex_BinaryLiteral = /^0b[01xX_']+$/
const regex_DecimalLiteral = /^[0-9_']+$/
const regex_RangeLiteral = /^[0-9]+\.\.[0-9]+$/
const regex_Punctuator = /[!%\^&\*\(\)\-\+={}\|~\[\];:"<>\?,\.#\/\\]/
const regex_Whitespace = /\s/

function IsKeyword(str: string): boolean {
    return s_Identifiers.includes(str)
}

export class TokenizerError extends Error {
    constructor(msg: string, public location: SourceLocation, cause: string | null = null) {
        super(msg)
        this.name = "TokenizerError"
        if (cause)
            this.cause = cause
    }
}

const _TOKENIZER_LOG_ENABLED = true
function TOKENIZER_LOG(...args: any) {
    if (!_TOKENIZER_LOG_ENABLED)
        return
    console.log("[Tokenizer] ", ...args)
}

export class Tokenizer {
    m_Tokens: Token[] = []
    constructor() {
    }

    Tokenize(content: string, sourceLocation: SourceLocation) {
        TOKENIZER_LOG(`Process source "${sourceLocation.file_name}" (${sourceLocation.file_path})`)

        const SEGMENT_VALUE = 0
        const SEGMENT_TYPE = 1
        const SEGMENT_LOCATION = 2

        // end padding for single position lookahead and remove carriage returns
        content = content.replace(/\r/g, "") + " "
        const contentLen = content.length

        var currentValue = ""
        var tokenActive = false
        var tokenStartIndex = 0
        var tokenStartLine = 0
        var tokenStartColumn = 0
        var segments: Array<Token> = []
        var commentActive = false
        var commentIsMultiline = false
        var stringActive = false

        var tokenPos = 0
        var lineNumber = 1
        var linePos = 0
        var lastLinePos = 0

        if (!regex_Whitespace.test(content.charAt(0))) {
            tokenActive = true
            tokenStartIndex = 0
            tokenStartLine = 1
            tokenStartColumn = 0
            linePos = 1
            tokenPos = 1
        }

        for (var i = 0; i < contentLen - 1; i++) {
            const c = content.charAt(i)

            tokenPos++
            if (c === "\n") {
                lastLinePos = linePos
                linePos = 1
                lineNumber++
            } else {
                linePos++
            }

            if (tokenActive) {
                if (!commentActive && !stringActive) {
                    const nextChar = content.charAt(i + 1)
                    const is_comment_start = (c == "/" && (nextChar == "/" || nextChar == "*"))
                    const is_string_start = c == '"'

                    if (is_comment_start) {
                        commentIsMultiline = nextChar == "*"
                        commentActive = true
                        tokenStartIndex = tokenPos - 1
                        tokenStartLine = lineNumber
                        tokenStartColumn = linePos - 1
                    } else if (is_string_start) {
                        stringActive = true
                        currentValue = c
                        tokenStartIndex = tokenPos - 1
                        tokenStartLine = lineNumber
                        tokenStartColumn = linePos - 1
                        continue
                    }
                }

                if (stringActive) {
                    const is_string_end = c == '"'
                    currentValue += c
                    if (is_string_end) {
                        segments.push(<Token>{
                            value: currentValue,
                            type: TokenType.LITERAL,
                            entry_size: tokenPos - tokenStartIndex - 1,
                            line: tokenStartLine,
                            column: tokenStartColumn
                        })
                        currentValue = ""
                        tokenActive = false
                        stringActive = false
                    }
                } else {
                    if (regex_Whitespace.test(c) && !commentActive) {
                        segments.push(<Token>{
                            value: currentValue,
                            type: IsKeyword(currentValue) ? TokenType.KEYWORD : TokenType.IDENTIFIER,
                            entry_size: tokenPos - tokenStartIndex - 1,
                            line: tokenStartLine,
                            column: tokenStartColumn
                        })
                        tokenActive = false
                        currentValue = ""
                    } else if (regex_Punctuator.test(c) && !commentActive) {
                        if (currentValue !== "") {
                            segments.push(<Token>{
                                value: currentValue,
                                type: IsKeyword(currentValue) ? TokenType.KEYWORD : TokenType.IDENTIFIER,
                                entry_size: tokenPos - (tokenStartIndex - 1),
                                line: tokenStartLine,
                                column: tokenStartColumn - 1
                            })
                            segments.push(<Token>{
                                value: c,
                                type: TokenType.PUNCTUATOR,
                                entry_size: 1,
                                line: lineNumber,
                                column: linePos - 1
                            })
                            currentValue = ""
                        } else {
                            segments.push(<Token>{
                                value: c, type: TokenType.PUNCTUATOR,
                                entry_size: 1,
                                line: lineNumber,
                                column: linePos - 1
                            })
                        }
                        tokenActive = false
                    } else {
                        const is_multiline_end = commentActive && (c == "/" && content.charAt(i - 1) == "*")
                        if (commentActive && ((!commentIsMultiline && c == "\n") || (commentActive && commentIsMultiline && is_multiline_end))) {
                            commentActive = false
                            tokenActive = false
                            segments.push(<Token>{
                                value: commentIsMultiline ? currentValue + "/" : currentValue,
                                type: TokenType.COMMENT,
                                entry_size: tokenPos - tokenStartIndex - 1,
                                line: tokenStartLine,
                                column: tokenStartColumn
                            })
                            currentValue = ""
                        } else {
                            currentValue += c
                        }
                    }
                }
            } else {
                if (regex_Whitespace.test(c)) {
                    continue
                } else {
                    tokenActive = true
                    tokenStartIndex = tokenPos
                    tokenStartLine = lineNumber
                    tokenStartColumn = linePos

                    i--
                    tokenPos--
                    if (c === "\n") {
                        linePos--
                        lineNumber--
                    } else {
                        linePos--
                    }
                }
            }
        }

        if (tokenActive) {
            if (commentActive) {
                segments.push(<Token>{
                    value: commentIsMultiline ? currentValue + "/" : currentValue,
                    type: TokenType.COMMENT,
                    entry_size: tokenPos - tokenStartIndex,
                    line: tokenStartLine,
                    column: tokenStartColumn
                })
            } else if (stringActive) {
                if (currentValue.charAt(currentValue.length - 1) == '"') {
                    segments.push(<Token>{
                        value: currentValue,
                        type: TokenType.LITERAL,
                        entry_size: tokenPos - tokenStartIndex,
                        line: tokenStartLine,
                        column: tokenStartColumn
                    })
                } else {
                    segments.push(<Token>{
                        value: currentValue,
                        type: TokenType.INVALID_LITERAL,
                        entry_size: tokenPos - tokenStartIndex,
                        line: tokenStartLine,
                        column: tokenStartColumn
                    })
                }
            } else {
                segments.push(<Token>{
                    value: currentValue,
                    type: IsKeyword(currentValue) ? TokenType.KEYWORD : TokenType.IDENTIFIER,
                    entry_size: tokenPos - tokenStartIndex,
                    line: tokenStartLine,
                    column: tokenStartColumn
                })
            }
        }

        for (var i = 0; i < segments.length; i++) {
            const seg = segments[i]
            if (seg.type == TokenType.IDENTIFIER) {
                if (!regex_Literal.test(seg.value)) {
                    if (regex_HexLiteral.test(seg.value)) {
                        segments[i].type = TokenType.LITERAL
                    } else if (regex_BinaryLiteral.test(seg.value)) {
                        segments[i].type = TokenType.LITERAL
                    } else if (regex_DecimalLiteral.test(seg.value)) {
                        segments[i].type = TokenType.LITERAL
                    } else if (regex_RangeLiteral.test(seg.value)) {
                        segments[i].type = TokenType.LITERAL
                    } else {
                        segments[i].type = TokenType.INVALID_IDENTIFIER
                        throw new TokenizerError(`Invalid identifier: '${seg.value}'`, new SourceLocation({ file_name: sourceLocation.file_name, file_path: sourceLocation.file_path, line: seg.line, column: seg.column, entry_size: seg.entry_size }))
                    }
                }
            }
        }

        var segIdx = 1
        for (const s of segments) {
            console.log(chalk.blueBright(`[Segment] (${segIdx++}/${segments.length}):`) +
                `\n` + chalk.blueBright("    |") + chalk.gray(` '${s.value}'`) +
                `\n` + chalk.blueBright("    |") + chalk.greenBright(` '${TokenType[s.type]}'`) +
                `\n` + chalk.blueBright("    |") + chalk.greenBright(` '${s.entry_size}'`) +
                `\n` + chalk.blueBright("    |") + chalk.yellowBright(` "${sourceLocation.file_path}:${s.line}:${s.column}"`))
        }
    }
}

File: XRT/src/Language/Tokenizer.cpp
#include "Tokenizer.hpp"
#include <Log/Logger.hpp>
#include <map>
#include <vector>
#include <algorithm>
#include <StringUtils.hpp>
#include <regex/ctre.hpp>
#include <ScopeExecTime.hpp>

#define REGEX_LITERAL         L"^[a-zA-Z_$][a-zA-Z_$0-9]*$"
#define REGEX_HEX_LITERAL     L"^0x[0-9a-fA-FxX_']+$"
#define REGEX_BINARY_LITERAL  L"^0b[01xX_']+$"
#define REGEX_DECIMAL_LITERAL L"^[0-9_']+$"
#define REGEX_RANGE_LITERAL   L"^[0-9]+\\.\\.[0-9]+$"
#define REGEX_PUNCTUATOR      L"[!%\\^&\\*\\(\\)\\-\\+={}\\|~\\[\\];:\"<>\?,\\.#\\/\\\\]"
#define REGEX_WHITESPACE      L"\\s"

namespace XRT {

    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    static std::vector<std::wstring_view> s_Identifiers = {
        L"namespace",
        L"component",
        L"ports",
        L"registers",
        L"implementation",
        L"extern",
        L"in",
        L"out",
        L"inout",
        L"if",
        L"else",
        L"for",
        L"template",
        L"typename",
        L"include",
        L"using",
        L"constexpr",
    };
    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    static bool IsKeyword(const std::wstring_view& str) {
        return std::find(s_Identifiers.begin(), s_Identifiers.end(), str) != s_Identifiers.end();
    }

    void Tokenizer::Tokenize(const std::string& _source, std::filesystem::path* filePath) {
        ScopeExecTime xt("Tokenize");

        std::wstring source = StringUtils::utf8_to_utf16(_source + " ");
        source.erase(std::remove(source.begin(), source.end(), (wchar_t)'\r'), source.end());

        std::vector<Token*> tokens;
        tokens.reserve(128);
        std::wstring currentValue;
        currentValue.reserve(128);

        bool tokenActive        = false;
        size_t tokenStartIndex  = 0;
        size_t tokenStartLine   = 0;
        size_t tokenStartColumn = 0;
        bool commentActive      = false;
        bool commentIsMultiline = false;
        bool stringActive       = false;
        size_t tokenPos         = 0;
        size_t lineNumber       = 1;
        size_t lineOffset       = 0;
        size_t lastLinePos      = 0;

        if (!ctre::starts_with<REGEX_WHITESPACE>(source.begin(), source.begin() + 1)) {
            tokenActive      = true;
            tokenStartIndex  = 0;
            tokenStartLine   = 1;
            tokenStartColumn = 0;
            lineOffset       = 1;
            tokenPos         = 1;
        }

        // Source is padded with 1 space at the end
        auto sourceLength              = source.length();
        std::wstring::iterator it_char = source.begin();
        for (auto i = 0; i < sourceLength - 1; i++) {
            auto c = *it_char;

            tokenPos++;
            if (c == '\n') {
                lastLinePos = lineOffset;
                lineOffset  = 1;
                lineNumber++;
            } else {
                lineOffset++;
            }

            if (tokenActive) {
                if (!commentActive & !stringActive) {
                    auto nextChar         = *(it_char + 1);
                    auto is_comment_start = (c == '/' && (nextChar == '/' || nextChar == '*'));
                    auto is_string_start  = c == '"';

                    if (is_comment_start) {
                        commentIsMultiline = nextChar == '*';
                        commentActive      = true;
                        tokenStartIndex    = tokenPos - 1;
                        tokenStartLine     = lineNumber;
                        tokenStartColumn   = lineOffset - 1;
                    } else if (is_string_start) {
                        stringActive     = true;
                        currentValue     = c;
                        tokenStartIndex  = tokenPos - 1;
                        tokenStartLine   = lineNumber;
                        tokenStartColumn = lineOffset - 1;
                        it_char++;
                        continue;
                    }
                }

                if (stringActive) {
                    auto is_string_end = c == '"';
                    currentValue += c;
                    if (is_string_end) {
                        tokens.push_back(new Token(
                            currentValue, TokenType::LITERAL, tokenPos - tokenStartIndex - 1, tokenStartLine, tokenStartColumn, filePath));
                        currentValue.clear();
                        tokenActive  = false;
                        stringActive = false;
                    }
                } else {
                    if (ctre::starts_with<REGEX_WHITESPACE>(it_char, it_char + 1) && !commentActive) {
                        tokens.push_back(new Token(currentValue,
                                                   IsKeyword(currentValue) ? TokenType::KEYWORD : TokenType::IDENTIFIER,
                                                   tokenPos - tokenStartIndex - 1,
                                                   tokenStartLine,
                                                   tokenStartColumn,
                                                   filePath));
                        tokenActive = false;
                        currentValue.clear();
                    } else if (ctre::starts_with<REGEX_PUNCTUATOR>(it_char, it_char + 1) && !commentActive) {
                        if (!currentValue.empty()) {
                            tokens.push_back(new Token(currentValue,
                                                       IsKeyword(currentValue) ? TokenType::KEYWORD : TokenType::IDENTIFIER,
                                                       tokenPos - (tokenStartIndex - 1),
                                                       tokenStartLine,
                                                       tokenStartColumn - 1,
                                                       filePath));
                            std::wstring cc;
                            cc = c;
                            tokens.push_back(new Token(cc, TokenType::PUNCTUATOR, 1, lineNumber, lineOffset - 1, filePath));
                            currentValue.clear();
                        } else {
                            std::wstring cc;
                            cc = c;
                            tokens.push_back(new Token(cc, TokenType::PUNCTUATOR, 1, lineNumber, lineOffset - 1, filePath));
                        }
                        tokenActive = false;
                    } else {
                        auto is_multiline_end = commentActive && (c == '/' && (i ? *(it_char - 1) : (wchar_t)'\0') == '*');
                        if (commentActive &&
                            ((!commentIsMultiline && c == '\n') || (commentActive && commentIsMultiline && is_multiline_end))) {
                            commentActive = false;
                            tokenActive   = false;
                            tokens.push_back(new Token(commentIsMultiline ? currentValue + L"/" : currentValue,
                                                       TokenType::COMMENT,
                                                       tokenPos - tokenStartIndex - 1,
                                                       tokenStartLine,
                                                       tokenStartColumn,
                                                       filePath));
                            currentValue.clear();
                        } else {
                            currentValue += c;
                        }
                    }
                }
            } else {
                if (!ctre::starts_with<REGEX_WHITESPACE>(it_char, it_char + 1)) {
                    tokenActive      = true;
                    tokenStartIndex  = tokenPos;
                    tokenStartLine   = lineNumber;
                    tokenStartColumn = lineOffset;

                    i--;
                    it_char--;
                    tokenPos--;
                    lineOffset--;
                    if (c == '\n') {
                        lineNumber--;
                    }
                }
            }

            it_char++;
        }
    }

} // namespace XRT